### ISSUE TITLE
Scan Benchmarks

### ISSUE THREAD

**Issue by pombredanne** [nexB/scancode-toolkit#378]
Following on the discussion in https://github.com/nexB/scancode-toolkit/pull/376  we should have and run scan benchmarks.

 @pombredanne commented:

>    We should create some benchmarks to watch for any perf changes btw.

 @sschuberth commented:
> That would be nice to have indeed, i.e. to have tests to identify performance regressions (or improvements).

>As a first step, looking at the times Travis CI builds take could be useful. Of course, this includes the setup and build times while we're only interested in the time to run tests, and the tests might be too many small projects whereas scanning performance probably is better measured with a single big project, but still.

>So I was quickly giving http://scribu.github.io/travis-stats/#nexB/scancode-toolkit a look, but unfortunately it does not render properly for me. See scribu/travis-stats#12.

**Comment by sschuberth** [comment]
FYI, https://scribu.github.io/travis-stats/#nexB/scancode-toolkit/develop looks much better now after @scribu was very quick at fixing some isses.

**Comment by pombredanne** [comment]
Very nice. Not sure which story this tells yet, but it looks quite decent.

**Comment by pombredanne** [comment]
The goal is to create a benchmark between scancode and other similar tools.
1. on license detection first
2. on copyright detection second

I have listed some of the tools to consider here: https://wiki.debian.org/CopyrightReviewTools

Some of the things to consider:
1. how to create a comprehensive test data set? How to get feedback on it? how to allow the community to add new tests? (web app?, posts on license-discuss, spdx? )
2. how to set the ground truth of what a perfect detection would be?
3. what would be the metrics to use (precision? recall? F1?)
4. how and where to run the benchmarks?
5. which tools to consider in the benchmarks?
6. how to map the results of each tool to the standard expectations? (e.g different ways of naming licenses, etc)
7. how to publish and publicize the results of the benchmarks? automated web app?
8. how to involve these other tools projects in the effort?
9. how to make the benchmark as serious, unbiased, scientific and factual as possible?
10.  how to keep things up to date and rerun the benchmark on a regular basis?
11. which os top run the benchs on? IMHO Linux only. case solved.

**Comment by pombredanne** [comment]
FYI with the merge of #885 in develop here are some tidbits. The license scan enjoys a ~ 20% speedup!
Not too bad.

```
Both are using a Linux kernel 4.1.5 rc2 from https://git.kernel.org/torvalds/t/linux-4.15-rc2.tar.gz

###############################################
With develop, 2180-02-07
###############################################

$ date;scancode -l -n 6 linux-4.15-rc2 -f json linux-4.15-rc2-develop-2018-02-07-notext.json;date
Wed Feb  7 12:39:26 CET 2018
Scanning files for: licenses with 6 process(es)...
Building license detection index...Done.
Scanning files...
[####################] 66423                                                           
Scanning done.
Scan statistics: 66423 files scanned in 1308s.
Scan options:    licenses with 6 process(es).
Scanning speed:  50.93 files per sec.
Scanning time:   1304s.
Indexing time:   4s.
Saving results.
Wed Feb  7 13:01:35 CET 2018


###############################################
with develop after #885 merge using the memory only (the on-disk caching is only mildly slower)
###############################################

$ scancode -l --max-in-memory=0  -n 6 linux-4.15-rc2 --json linux-4.15-rc2-b787-latest-notext-all-ram.json 
Setup plugins...
Collect file inventory...
Scan files for: licenses with 6 process(es)...
[####################] 62047                                                           
Scanning done.
Summary:        licenses with 6 process(es)
Errors count:   0
Scan Speed:     60.94 files/sec. 
Initial counts: 66424 resource(s): 62047 file(s) and 4377 directorie(s) 
Final counts:   66424 resource(s): 62047 file(s) and 4377 directorie(s) 
Timings:
  setup_scan:licenses: 2.33s
  setup: 2.33s
  inventory: 9.08s
  scan: 1018.13s
  output:json: 5.92s
  output: 5.92s
  total: 1036.93s

```

**Comment by pombredanne** [comment]
We started publishing regular results and high level data for scans in ScanCode.io at https://github.com/nexB/scancode.io-reference-scans
